cps/kni.c

```diff
@@ -1738,13 +1740,117 @@ loaded_kmod_matches_file(void *file, unsigned long len)
 	return false;
 }
 
+static int
+find_kni_kmod_path(char *path, size_t path_len, const char *alias)
+{
+	struct kmod_ctx *ctx;
+	char dirname[PATH_MAX];
+	struct utsname u;
+	struct kmod_list *l, *filtered, *list = NULL;
+	int ret;
+
+	/* Get kernel name and build module path. */
+	if (uname(&u) < 0) {
+		int saved_errno = errno;
+		CPS_LOG(ERR, "uname: %s\n", strerror(errno));
+		return -saved_errno;
+	}
+	ret = snprintf(dirname, sizeof(dirname),
+		"/lib/modules/%s", u.release);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(dirname));
+
+	/*
+	 * Create kmod library context using default configuration.
+	 * The context that's created has an inital refcount of 1.
+	 */
+	ctx = kmod_new(dirname, NULL);
+	if (ctx == NULL) {
+		CPS_LOG(ERR, "kmod_new failed\n");
+		return -ENOMEM;
+	}
+
+	/*
+	 * Create a list of kernel modules that match the alias.
+	 * The initial refcount of the list is 1 and must be released.
+	 */
+	ret = kmod_module_new_from_lookup(ctx, alias, &list);
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to lookup module alias %s\n", alias);
+		ret = -ENOMEM;
+		goto put_ctx;
+	}
+	if (list == NULL) {
+		CPS_LOG(ERR, "Module %s not found\n", alias);
+		ret = -ENOENT;
+		goto put_ctx;
+	}
+
+	/* Filter out builtin modules from the list.*/
+	ret = kmod_module_apply_filter(ctx, KMOD_FILTER_BUILTIN,
+		list, &filtered);
+
+	/* Filtered list is now stored in @filtered, so release @list. */
+	kmod_module_unref_list(list);
+
+	if (ret < 0) {
+		CPS_LOG(ERR, "Failed to filter kernel module list to find %s\n",
+			alias);
+		ret = -EINVAL;
+		goto put_ctx;
+	}
+	if (filtered == NULL) {
+		CPS_LOG(ERR, "Module %s not found\n", alias);
+		ret = -ENOENT;
+		goto put_ctx;
+	}
+
+	kmod_list_foreach(l, filtered) {
+		struct kmod_module *mod = kmod_module_get_module(l);
+		const char *kmod_name = kmod_module_get_name(mod);
+		const char *kmod_path;
+
+		/* Not the module we're looking for. */
+		if (strcmp(kmod_name, alias)) {
+			kmod_module_unref(mod);
+			continue;
+		}
+
+		kmod_path = kmod_module_get_path(mod);
+		if (strlen(kmod_path) > path_len - 1) {
+			CPS_LOG(ERR, "Found kernel module path (%s) but buffer is too short to hold it\n",
+				kmod_path);
+			ret = -ENOMEM;
+		} else
+			strcpy(path, kmod_path);
```

Comment:
At this point, we know that `ret >= 0`, but it's better to make sure it's `ret = 0;`.